network/peers/connections/handshaker.go

```diff
@@ -186,8 +199,39 @@ func (h *handshaker) Handshake(logger *zap.Logger, conn libp2pnetwork.Conn) erro
 	if ni == nil {
 		return errors.New("empty node info")
 	}
+
+	privateKey, found, err := h.nodeStorage.GetPrivateKey()
+	if !found {
+		return err
+	}
+
+	publicKey, err := rsaencryption.ExtractPublicKeyPem(privateKey)
+	if err != nil {
+		return err
+	}
+
+	handshakeData := records.HandshakeData{
+		SenderPeerID:    h.net.LocalPeer(),
+		RecipientPeerID: pid,
+		Timestamp:       time.Now(),
+		SenderPubKeyPem: publicKey,
+	}
+	hash := handshakeData.Hash()
+
+	signature, err := rsa.SignPKCS1v15(nil, privateKey, crypto.SHA256, hash[:])
+	if err != nil {
+		logger.Debug("could not sign", zap.Error(err))
+		return err
+	}
+
+	sni := records.SignedNodeInfo{
+		NodeInfo:      ni,
```

Comment:
I'm confused, `ni` is received from `nodeInfoFromStream` it means that it the remote peer's NodeInfo, how come we sign on it? we should verify it and create our own NodeInfo to sign on.

Reply:
fixed!